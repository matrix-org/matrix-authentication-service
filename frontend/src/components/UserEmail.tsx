// Copyright 2023 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {
  Button,
  Control,
  Field,
  Label,
  Message,
  Root as Form,
  Submit,
} from "@vector-im/compound-web";
import { atom, useAtom, useSetAtom } from "jotai";
import { atomFamily } from "jotai/utils";
import { atomWithMutation } from "jotai-urql";
import { useTransition } from "react";

import { FragmentType, graphql, useFragment } from "../gql";

import Block from "./Block";
import DateTime from "./DateTime";
import Typography from "./Typography";

// This component shows a single user email address, with controls to verify it,
// resend the verification email, remove it, and set it as the primary email address.

const FRAGMENT = graphql(/* GraphQL */ `
  fragment UserEmail_email on UserEmail {
    id
    email
    createdAt
    confirmedAt
  }
`);

const VERIFY_EMAIL_MUTATION = graphql(/* GraphQL */ `
  mutation VerifyEmail($id: ID!, $code: String!) {
    verifyEmail(input: { userEmailId: $id, code: $code }) {
      status

      user {
        id
        primaryEmail {
          id
        }
      }

      email {
        id
        ...UserEmail_email
      }
    }
  }
`);

const RESEND_VERIFICATION_EMAIL_MUTATION = graphql(/* GraphQL */ `
  mutation ResendVerificationEmail($id: ID!) {
    sendVerificationEmail(input: { userEmailId: $id }) {
      status

      user {
        id
        primaryEmail {
          id
        }
      }

      email {
        id
        ...UserEmail_email
      }
    }
  }
`);

const REMOVE_EMAIL_MUTATION = graphql(/* GraphQL */ `
  mutation RemoveEmail($id: ID!) {
    removeEmail(input: { userEmailId: $id }) {
      status

      user {
        id
      }
    }
  }
`);

const SET_PRIMARY_EMAIL_MUTATION = graphql(/* GraphQL */ `
  mutation SetPrimaryEmail($id: ID!) {
    setPrimaryEmail(input: { userEmailId: $id }) {
      status
      user {
        id
        primaryEmail {
          id
        }
      }
    }
  }
`);

const verifyEmailFamily = atomFamily((id: string) => {
  const verifyEmail = atomWithMutation(VERIFY_EMAIL_MUTATION);

  // A proxy atom which pre-sets the id variable in the mutation
  const verifyEmailAtom = atom(
    (get) => get(verifyEmail),
    (get, set, code: string) => set(verifyEmail, { id, code })
  );

  return verifyEmailAtom;
});

const resendVerificationEmailFamily = atomFamily((id: string) => {
  const resendVerificationEmail = atomWithMutation(
    RESEND_VERIFICATION_EMAIL_MUTATION
  );

  // A proxy atom which pre-sets the id variable in the mutation
  const resendVerificationEmailAtom = atom(
    (get) => get(resendVerificationEmail),
    (get, set) => set(resendVerificationEmail, { id })
  );

  return resendVerificationEmailAtom;
});

const removeEmailFamily = atomFamily((id: string) => {
  const removeEmail = atomWithMutation(REMOVE_EMAIL_MUTATION);

  // A proxy atom which pre-sets the id variable in the mutation
  const removeEmailAtom = atom(
    (get) => get(removeEmail),
    (get, set) => set(removeEmail, { id })
  );

  return removeEmailAtom;
});

const setPrimaryEmailFamily = atomFamily((id: string) => {
  const setPrimaryEmail = atomWithMutation(SET_PRIMARY_EMAIL_MUTATION);

  // A proxy atom which pre-sets the id variable in the mutation
  const setPrimaryEmailAtom = atom(
    (get) => get(setPrimaryEmail),
    (get, set) => set(setPrimaryEmail, { id })
  );

  return setPrimaryEmailAtom;
});

const UserEmail: React.FC<{
  email: FragmentType<typeof FRAGMENT>;
  onRemove?: () => void;
  onSetPrimary?: () => void;
  isPrimary?: boolean;
  highlight?: boolean;
}> = ({ email, isPrimary, highlight, onSetPrimary, onRemove }) => {
  const [pending, startTransition] = useTransition();
  const data = useFragment(FRAGMENT, email);
  const [verifyEmailResult, verifyEmail] = useAtom(verifyEmailFamily(data.id));
  const [resendVerificationEmailResult, resendVerificationEmail] = useAtom(
    resendVerificationEmailFamily(data.id)
  );
  const setPrimaryEmail = useSetAtom(setPrimaryEmailFamily(data.id));
  const removeEmail = useSetAtom(removeEmailFamily(data.id));
  // TODO: compound doesn't forward the refs properly
  // const fieldRef = useRef<HTMLInputElement>(null);

  const onFormSubmit = (e: React.FormEvent<HTMLFormElement>): void => {
    e.preventDefault();
    const formData = new FormData(e.currentTarget);
    const code = formData.get("code") as string;
    startTransition(() => {
      verifyEmail(code).then((result) => {
        // Clear the form
        e.currentTarget?.reset();

        if (result.data?.verifyEmail.status === "VERIFIED") {
          // Call the onSetPrimary callback if provided
          // XXX: do we need a dedicated onVerify callback?
          onSetPrimary?.();
        }
      });
    });
  };

  const onResendClick = (): void => {
    startTransition(() => {
      resendVerificationEmail().then(() => {
        // TODO: fieldRef.current?.focus();
      });
    });
  };

  const onRemoveClick = (): void => {
    startTransition(() => {
      removeEmail().then(() => {
        // Call the onRemove callback if provided
        onRemove?.();
      });
    });
  };

  const onSetPrimaryClick = (): void => {
    startTransition(() => {
      setPrimaryEmail().then(() => {
        // Call the onSetPrimary callback if provided
        onSetPrimary?.();
      });
    });
  };

  const emailSent =
    resendVerificationEmailResult.data?.sendVerificationEmail.status === "SENT";
  const invalidCode =
    verifyEmailResult.data?.verifyEmail.status === "INVALID_CODE";

  return (
    <Block
      highlight={highlight}
      className="grid grid-col-1 gap-2 pb-4 border-b-2 border-b-grey-200"
    >
      {isPrimary && (
        <Typography variant="body" bold>
          Primary
        </Typography>
      )}
      <Typography variant="caption" bold className="flex-1">
        {data.email}
      </Typography>
      {data.confirmedAt ? (
        <Typography variant="micro">
          Verified <DateTime datetime={data.confirmedAt} />
        </Typography>
      ) : (
        <Form onSubmit={onFormSubmit} className="grid grid-cols-2 gap-2">
          <Field name="code" className="col-span-2">
            <Label>Code</Label>
            <Control
              // ref={fieldRef}
              placeholder="xxxxxx"
              type="text"
              inputMode="numeric"
            />
          </Field>
          {invalidCode && (
            <Message className="col-span-2 text-alert font-bold">
              Invalid code
            </Message>
          )}
          <Submit size="sm" type="submit" disabled={pending}>
            Submit
          </Submit>
          <Button
            size="sm"
            kind="secondary"
            disabled={pending || emailSent}
            onClick={onResendClick}
          >
            {emailSent ? "Sent!" : "Resend"}
          </Button>
        </Form>
      )}
      {!isPrimary && (
        <div className="flex justify-between items-center">
          {/* The primary email can only be set if the email was verified */}
          {data.confirmedAt ? (
            <Button size="sm" disabled={pending} onClick={onSetPrimaryClick}>
              Set primary
            </Button>
          ) : (
            <div />
          )}
          <Button
            kind="destructive"
            size="sm"
            disabled={pending}
            onClick={onRemoveClick}
          >
            Remove
          </Button>
        </div>
      )}
    </Block>
  );
};

export default UserEmail;
