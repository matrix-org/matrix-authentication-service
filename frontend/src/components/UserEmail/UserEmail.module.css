/* Copyright 2023 The Matrix.org Foundation C.I.C.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

.user-email {
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  gap: var(--cpd-space-2x);
}

.user-email-line {
  display: flex;
  align-self: stretch;
  gap: var(--cpd-space-2x);
}

.user-email-delete-icon {
  color: var(--cpd-color-icon-critical-primary);
}

button[disabled] .user-email-delete-icon {
  color: var(--cpd-color-icon-disabled);
}

.user-email-field {
  flex: 1;
}

.user-email-unverified {
  color: var(--cpd-color-text-critical-primary);
}

.link {
  display: inline-block;
  text-decoration: underline;
  color: var(--cpd-color-text-primary);
  font-weight: var(--cpd-font-weight-medium);
  border-radius: var(--cpd-radius-pill-effect);
  padding-inline: 0.25rem;
}

.link:hover {
  background: var(--cpd-color-gray-300);
}

.link:active {
  color: var(--cpd-color-text-on-solid-primary);
}
