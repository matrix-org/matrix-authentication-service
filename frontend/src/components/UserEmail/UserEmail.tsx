// Copyright 2023 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import IconDelete from "@vector-im/compound-design-tokens/icons/delete.svg?react";
import { Form, IconButton, Text, Tooltip } from "@vector-im/compound-web";
import { atom, useSetAtom } from "jotai";
import { atomFamily } from "jotai/utils";
import { atomWithMutation } from "jotai-urql";
import { useTransition, ComponentProps, ReactNode } from "react";
import { Translation, useTranslation } from "react-i18next";

import { FragmentType, graphql, useFragment } from "../../gql";
import { Link } from "../../routing";
import ConfirmationModal from "../ConfirmationModal/ConfirmationModal";

import styles from "./UserEmail.module.css";

// This component shows a single user email address, with controls to verify it,
// resend the verification email, remove it, and set it as the primary email address.

export const FRAGMENT = graphql(/* GraphQL */ `
  fragment UserEmail_email on UserEmail {
    id
    email
    confirmedAt
  }
`);

const REMOVE_EMAIL_MUTATION = graphql(/* GraphQL */ `
  mutation RemoveEmail($id: ID!) {
    removeEmail(input: { userEmailId: $id }) {
      status

      user {
        id
      }
    }
  }
`);

const SET_PRIMARY_EMAIL_MUTATION = graphql(/* GraphQL */ `
  mutation SetPrimaryEmail($id: ID!) {
    setPrimaryEmail(input: { userEmailId: $id }) {
      status
      user {
        id
        primaryEmail {
          id
        }
      }
    }
  }
`);

const removeEmailFamily = atomFamily((id: string) => {
  const removeEmail = atomWithMutation(REMOVE_EMAIL_MUTATION);

  // A proxy atom which pre-sets the id variable in the mutation
  const removeEmailAtom = atom(
    (get) => get(removeEmail),
    (_get, set) => set(removeEmail, { id }),
  );

  return removeEmailAtom;
});

const setPrimaryEmailFamily = atomFamily((id: string) => {
  const setPrimaryEmail = atomWithMutation(SET_PRIMARY_EMAIL_MUTATION);

  // A proxy atom which pre-sets the id variable in the mutation
  const setPrimaryEmailAtom = atom(
    (get) => get(setPrimaryEmail),
    (_get, set) => set(setPrimaryEmail, { id }),
  );

  return setPrimaryEmailAtom;
});

const DeleteButton: React.FC<{ disabled?: boolean; onClick?: () => void }> = ({
  disabled,
  onClick,
}) => (
  <Translation>
    {(t): ReactNode => (
      <Tooltip label={t("frontend.user_email.delete_button_title")}>
        <IconButton
          type="button"
          disabled={disabled}
          className="m-2"
          onClick={onClick}
          size="var(--cpd-space-8x)"
        >
          <IconDelete className={styles.userEmailDeleteIcon} />
        </IconButton>
      </Tooltip>
    )}
  </Translation>
);

const DeleteButtonWithConfirmation: React.FC<
  ComponentProps<typeof DeleteButton>
> = ({ onClick, ...rest }) => {
  const { t } = useTranslation();
  const onConfirm = (): void => {
    onClick?.();
  };

  // NOOP function, otherwise we dont render a cancel button
  const onDeny = (): void => {};

  return (
    <>
      <ConfirmationModal
        trigger={<DeleteButton {...rest} />}
        onDeny={onDeny}
        onConfirm={onConfirm}
      >
        <Text>
          {t("frontend.user_email.delete_button_confirmation_modal.body")}
        </Text>
      </ConfirmationModal>
    </>
  );
};

const UserEmail: React.FC<{
  email: FragmentType<typeof FRAGMENT>;
  onRemove?: () => void;
  onSetPrimary?: () => void;
  isPrimary?: boolean;
  highlight?: boolean;
}> = ({ email, isPrimary, highlight, onSetPrimary, onRemove }) => {
  const [pending, startTransition] = useTransition();
  const data = useFragment(FRAGMENT, email);
  const setPrimaryEmail = useSetAtom(setPrimaryEmailFamily(data.id));
  const removeEmail = useSetAtom(removeEmailFamily(data.id));
  const { t } = useTranslation();

  const onRemoveClick = (): void => {
    startTransition(() => {
      removeEmail().then(() => {
        // Call the onRemove callback if provided
        onRemove?.();
      });
    });
  };

  const onSetPrimaryClick = (): void => {
    startTransition(() => {
      setPrimaryEmail().then(() => {
        // Call the onSetPrimary callback if provided
        onSetPrimary?.();
      });
    });
  };

  return (
    <Form.Root>
      <Form.Field name="email">
        <Form.Label>
          {isPrimary
            ? t("frontend.user_email.primary_email")
            : t("frontend.user_email.email")}
        </Form.Label>
        <div className="flex">
          <Form.TextControl
            type="email"
            readOnly
            value={data.email}
            className={styles.userEmailField}
          />
          <DeleteButtonWithConfirmation
            disabled={isPrimary || pending}
            onClick={onRemoveClick}
          />
        </div>

        <Form.HelpMessage>
          {data.confirmedAt && !isPrimary && (
            <button
              className={styles.link}
              disabled={pending}
              onClick={onSetPrimaryClick}
            >
              {t("frontend.user_email.make_primary_button")}
            </button>
          )}
          {!data.confirmedAt && (
            <>
              <span className={styles.userEmailUnverified}>
                {t("frontend.user_email.unverified")}
              </span>{" "}
              |{" "}
              <Link kind="button" route={{ type: "verify-email", id: data.id }}>
                {t("frontend.user_email.retry_button")}
              </Link>
            </>
          )}
        </Form.HelpMessage>
      </Form.Field>
    </Form.Root>
  );
};

export default UserEmail;
