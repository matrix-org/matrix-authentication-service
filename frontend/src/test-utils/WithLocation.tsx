// Copyright 2023 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// @vitest-environment happy-dom

import { Provider } from "jotai";
import { useHydrateAtoms } from "jotai/utils";

import { appConfigAtom, locationAtom } from "../routing";

const HydrateLocation: React.FC<React.PropsWithChildren<{ path: string }>> = ({
  children,
  path,
}) => {
  useHydrateAtoms([
    [appConfigAtom, { root: "/" }],
    [locationAtom, { pathname: path }],
  ]);
  return <>{children}</>;
};

/**
 * Utility for testing components that rely on routing or location
 * For example any component that includes a <Link />
 * Eg:
 * ```
 * const component = create(
      <WithLocation path="/">
        <NavItem route={{ type: "profile" }}>Active</NavItem>
      </WithLocation>,
    );
 * ```
 */
export const WithLocation: React.FC<
  React.PropsWithChildren<{ path?: string }>
> = ({ children, path }) => {
  return (
    <Provider>
      <HydrateLocation path={path || "/"}>{children}</HydrateLocation>
    </Provider>
  );
};
