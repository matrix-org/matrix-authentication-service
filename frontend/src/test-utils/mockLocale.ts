// Copyright 2023 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { vi } from "vitest";

/**
 * Mock the locale on Intl.DateTimeFormat
 * To achieve stable formatted dates across environments
 * Defaults to `en-GB`
 */
export const mockLocale = (defaultLocale = "en-GB"): void => {
  const { DateTimeFormat } = Intl;
  vi.spyOn(Intl, "DateTimeFormat").mockImplementation(
    (
      locales?: Intl.LocalesArgument,
      options?: Intl.DateTimeFormatOptions | undefined,
    ) => new DateTimeFormat(locales || defaultLocale, options),
  );
};
