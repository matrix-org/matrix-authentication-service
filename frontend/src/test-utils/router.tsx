// Copyright 2024 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {
  RouterProvider,
  createMemoryHistory,
  createRootRoute,
  createRoute,
  createRouter,
} from "@tanstack/react-router";

const rootRoute = createRootRoute();
const index = createRoute({ getParentRoute: () => rootRoute, path: "/" });
rootRoute.addChildren([index]);

const router = createRouter({
  history: createMemoryHistory(),
  routeTree: rootRoute,
});

const routerReady = router.load();
// Because we also use this in the stories, we need to make sure we call this only in tests
if (import.meta.vitest) {
  // Make sure the router is ready before running any tests
  import("vitest").then(({ beforeAll }) =>
    beforeAll(async () => await routerReady),
  );
}

export const DummyRouter: React.FC<React.PropsWithChildren> = ({
  children,
}) => (
  /** @ts-expect-error: The router we pass doesn't match the "real" router, which is fine for tests */
  <RouterProvider router={router} defaultComponent={() => children} />
);
