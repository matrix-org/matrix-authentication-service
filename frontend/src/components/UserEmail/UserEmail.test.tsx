// Copyright 2023 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// @vitest-environment happy-dom

import { render, cleanup, screen } from "@testing-library/react";
import { describe, expect, it, afterEach } from "vitest";

import { makeFragmentData } from "../../gql/fragment-masking";
import { WithLocation } from "../../test-utils/WithLocation";

import UserEmail, { FRAGMENT } from "./UserEmail";

describe("<UserEmail />", () => {
  afterEach(cleanup);

  const baseEmail = {
    id: "testid",
    email: "ernie@sesame.st",
  };

  it("renders a basic email", () => {
    const email = makeFragmentData(baseEmail, FRAGMENT);
    const { container } = render(
      <WithLocation>
        <UserEmail email={email} />
      </WithLocation>,
    );

    expect(container).toMatchSnapshot();
  });

  it("renders a primary email", () => {
    const email = makeFragmentData(baseEmail, FRAGMENT);
    render(
      <WithLocation>
        <UserEmail email={email} isPrimary={true} />
      </WithLocation>,
    );

    // uses primary email label
    expect(screen.getByText("Primary email")).toBeTruthy();
    // no make primary button
    expect(screen.queryByText("Make primary")).toBeFalsy();
    // cannot be removed
    expect(
      screen.getByLabelText("Remove email address").hasAttribute("disabled"),
    ).toBeTruthy();
  });

  it("renders a confirmed email", () => {
    const email = makeFragmentData(
      {
        ...baseEmail,
        confirmedAt: "2023-07-29T03:35:17.451292+00:00",
      },
      FRAGMENT,
    );
    render(
      <WithLocation>
        <UserEmail email={email} />
      </WithLocation>,
    );

    // has make primary button
    expect(screen.getByText("Make primary")).toBeTruthy();
  });

  describe("read only emails", () => {
    it("renders a basic email", () => {
      const email = makeFragmentData(baseEmail, FRAGMENT);
      render(
        <WithLocation>
          <UserEmail email={email} isReadOnly />
        </WithLocation>,
      );

      expect(screen.queryByLabelText("Remove email address")).toBeFalsy();
      // can still verify
      expect(screen.getByText("Retry verification")).toBeTruthy();
    });

    it("renders a primary email", () => {
      const email = makeFragmentData(baseEmail, FRAGMENT);
      render(
        <WithLocation>
          <UserEmail email={email} isReadOnly isPrimary={true} />
        </WithLocation>,
      );

      // cannot be removed
      expect(screen.queryByLabelText("Remove email address")).toBeFalsy();
    });

    it("renders a confirmed email", () => {
      const email = makeFragmentData(
        {
          ...baseEmail,
          confirmedAt: "2023-07-29T03:35:17.451292+00:00",
        },
        FRAGMENT,
      );
      render(
        <WithLocation>
          <UserEmail email={email} isReadOnly />
        </WithLocation>,
      );

      // no remove button
      expect(screen.queryByLabelText("Remove email address")).toBeFalsy();
      // no make primary button
      expect(screen.queryByText("Make primary")).toBeFalsy();
    });
  });
});
