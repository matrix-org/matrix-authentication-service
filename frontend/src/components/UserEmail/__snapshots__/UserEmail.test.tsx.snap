// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<UserEmail /> > renders a basic email 1`] = `
<div>
  <form
    class="_root_578ve_24"
  >
    <div
      class="_field_578ve_34"
    >
      <label
        class="_label_578ve_67"
        for="radix-:r0:"
      >
        Email
      </label>
      <div
        class="flex"
      >
        <input
          aria-describedby="radix-:r5:"
          class="_control_9gon8_18 _userEmailField_e2a518"
          id="radix-:r0:"
          name="email"
          readonly=""
          title=""
          type="email"
          value="ernie@sesame.st"
        />
        <button
          class="_icon-button_1qjaf_17 m-2"
          data-state="closed"
          style="--cpd-icon-button-size: var(--cpd-space-8x);"
          type="button"
        >
          <svg
            class="_userEmailDeleteIcon_e2a518"
            fill="none"
            height="1em"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M7 21C6.45 21 5.97917 20.8042 5.5875 20.4125C5.19583 20.0208 5 19.55 5 19V6C4.71667 6 4.47917 5.90417 4.2875 5.7125C4.09583 5.52083 4 5.28333 4 5C4 4.71667 4.09583 4.47917 4.2875 4.2875C4.47917 4.09583 4.71667 4 5 4H9C9 3.71667 9.09583 3.47917 9.2875 3.2875C9.47917 3.09583 9.71667 3 10 3H14C14.2833 3 14.5208 3.09583 14.7125 3.2875C14.9042 3.47917 15 3.71667 15 4H19C19.2833 4 19.5208 4.09583 19.7125 4.2875C19.9042 4.47917 20 4.71667 20 5C20 5.28333 19.9042 5.52083 19.7125 5.7125C19.5208 5.90417 19.2833 6 19 6V19C19 19.55 18.8042 20.0208 18.4125 20.4125C18.0208 20.8042 17.55 21 17 21H7ZM7 6V19H17V6H7ZM9 16C9 16.2833 9.09583 16.5208 9.2875 16.7125C9.47917 16.9042 9.71667 17 10 17C10.2833 17 10.5208 16.9042 10.7125 16.7125C10.9042 16.5208 11 16.2833 11 16V9C11 8.71667 10.9042 8.47917 10.7125 8.2875C10.5208 8.09583 10.2833 8 10 8C9.71667 8 9.47917 8.09583 9.2875 8.2875C9.09583 8.47917 9 8.71667 9 9V16ZM13 16C13 16.2833 13.0958 16.5208 13.2875 16.7125C13.4792 16.9042 13.7167 17 14 17C14.2833 17 14.5208 16.9042 14.7125 16.7125C14.9042 16.5208 15 16.2833 15 16V9C15 8.71667 14.9042 8.47917 14.7125 8.2875C14.5208 8.09583 14.2833 8 14 8C13.7167 8 13.4792 8.09583 13.2875 8.2875C13.0958 8.47917 13 8.71667 13 9V16Z"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
      <span
        class="_message_578ve_98 _help-message_578ve_104"
        id="radix-:r5:"
      >
        <span
          class="_userEmailUnverified_e2a518"
        >
          Unverified
        </span>
         
        |
         
        <a
          class="_linkButton_b80ad8"
          href="/emails/testid/verify"
        >
          Retry verification
        </a>
      </span>
    </div>
  </form>
</div>
`;
